### PR TITLE
fix(clip): preserve nulls when clipping

### LINKED ISSUE #7168: bug: .clip() doesn't preserve NULLs
### What happened?

I would expect nulls to stay nulls. But they get turned into the upper value (if upper bound is provided) or the lower bound (if no upper bound is provided)

```python
import ibis

t = ibis.memtable({"values": [0, 1.0, 2.0, 3.0, 4.0, None]})
t.mutate(clipped1=t.values.clip(lower=2.5, upper=3), clipped2=t.values.clip(lower=2.5))
┏━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┓
┃ values  ┃ clipped1 ┃ clipped2 ┃
┡━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━┩
│ float64 │ float64  │ float64  │
├─────────┼──────────┼──────────┤
│     0.0 │      2.5 │      2.5 │
│     1.0 │      2.5 │      2.5 │
│     2.0 │      2.5 │      2.5 │
│     3.0 │      3.0 │      3.0 │
│     4.0 │      3.0 │      4.0 │
│     nan │      3.0 │      2.5 │
└─────────┴──────────┴──────────┘
```

### What version of ibis are you using?

main

### What backend(s) are you using, if any?

duckdb, maybe this is irrelevant. Not sure how different backends handle NULLs in GREATEST() and LEAST()

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
Ensure that NULLs are propagated through `clip`

Fixes #7168.

### CODE REVIEW

File: ibis/expr/types/numeric.py [hunk 1]
```diff
@@ -198,7 +200,13 @@ def clip(
         --------
         >>> import ibis
         >>> ibis.options.interactive = True
-        >>> t = ibis.memtable({"values": range(8)})
+        >>> t = ibis.memtable(
```
Comment (NickCrews) [inline ibis/expr/types/numeric.py:203]: nit: can you just add some NULLs to the existing column of data? I think that shows all aspects of clip, and is less complicated and less lines

File: ibis/backends/tests/test_numeric.py [hunk 2]
```diff
@@ -1364,25 +1364,40 @@ def test_random(con):
...
-        (
+        param(
             lambda x: x.clip(lower=0, upper=1.0),
             lambda x: x.clip(lower=0, upper=1.0),
+            id="lower-upper-float",
+        ),
+        param(
+            lambda x: x.nullif(1).clip(lower=0),
+            lambda x: x.where(x != 1).clip(lower=0),
+            id="nullif-lower",
```
Comment (NickCrews) [inline ibis/backends/tests/test_numeric.py:1395]: nit: change id to "null-preserved-lower" or simply "null-lower", to show the intention of the test. "nullif" is more a description of how you accomplished that.

File: ibis/backends/pyspark/compiler.py [hunk 3]
```diff
@@ -775,11 +775,11 @@ def compile_clip(t, op, **kwargs):
 
     def column_min(value, limit):
         """Return values greater than or equal to `limit`."""
-        return F.when(value < limit, limit).otherwise(value)
+        return F.when((value < limit) & ~F.isnull(value), limit).otherwise(value)
```
Comment (NickCrews) [inline ibis/backends/pyspark/compiler.py:778]: Is this codecov warning accurate? It looks to me like this should get hit every time `compile_clip()` is called. Is `compile_clip()` not getting called at all?

**Review by NickCrews — APPROVED** [review]
few nits, but looks good, thanks!

File: ibis/backends/pyspark/compiler.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/pyspark/compiler.py:778]: It's probably not accurate.

File: ibis/backends/tests/test_numeric.py (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/tests/test_numeric.py]: Done.

**Pull request merged**